RedHatInsights/rhsm-subscriptions
# SWATCH-1238: Replace tag profile in hosts/instances APIs

Jira issue: [SWATCH-1238](https://issues.redhat.com/browse/SWATCH-1238)

Description
===========

Use swatch-product-configuration to lookup necessary information in the hosts and instances APIs, replacing tag profile in related classes.

Note that there is one messy aspect I had to handle which is commented in the code:

> NOTE: Measurement.Uom values use a convention of capitalized kebab-case (e.g.
Instance-hours). The DB values and our config both use a convention with uppercase snake-case (e.g. INSTANCE_HOURS). So in order to match them, we need to make them the same case and remove hyphens and underscores (e.g. instancehours).

Otherwise, the refactoring is fairly straight-forward.

Testing
=======

1. Deploy the service:

```shell
DEV_MODE=true ./gradlew :bootRun
```

2. Ensure opted-in:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/opt-in \
  x-rh-identity:$(echo -n '{"identity":{"account_number":"account123","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w0)
```

3. Invoke hosts API with a non-PAYG product:

```shell
http ":8000/api/rhsm-subscriptions/v1/hosts/products/RHEL for ARM" \
  x-rh-identity:$(echo -n '{"identity":{"account_number":"account123","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w0)
```

4. Invoke hosts API with a PAYG product:

```shell
http ":8000/api/rhsm-subscriptions/v1/hosts/products/rosa" \
  x-rh-identity:$(echo -n '{"identity":{"account_number":"account123","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w0)
```

5. Invoke instances API with a non-PAYG product:

```shell
http ":8000/api/rhsm-subscriptions/v1/instances/products/RHEL for ARM" \
  x-rh-identity:$(echo -n '{"identity":{"account_number":"account123","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w0)
```

6. Invoke instances API with a PAYG product:

```shell
http ":8000/api/rhsm-subscriptions/v1/instances/products/rosa" \
  x-rh-identity:$(echo -n '{"identity":{"account_number":"account123","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w0)
```

Verification
------------

No errors should be seen in the logs for any of the above calls.
